Inverse kinematic

AndreaCatania opened the pull request:
This is a short video that show IK in action https://youtu.be/hr2x5zWY_AQ

The code should be polished and the main thing to add are the constraints.
The constraints addition is not so important because the algorithm that I've used is FABRIK, and it doesn't need to be constrained to look "reasonably" and this mean that use it will be super easy.

In order to use it you don't need to attach a physical bone to the skeleton. You will need to add it only to apply a constraint (That it's not yer implemented).

## How to use it: 
To use it the first thing to do is create the IK task using this line of code:
```
var goal_transform = Transform()
var right_arm_chain = $Skeleton.create_ik_task("right_arm_bone", "right_hand_bone", goal_transform )
```

Then in the process function you have to call the function to solve the chain:
```
func _process(delta):
	$Skeleton.solve_ik_task( right_arm_chain );
```

Only if the position of task changes update it using:
```
var goal_transform = Transform()
$Skeleton.set_ik_task_goal(right_arm_chain , goal_transform )
```

AndreaCatania committed:
Implemented IK

AndreaCatania commented:
The main idea is that provide an API that allow us to use IK on skeleton, then implement some function (or node I hope) allow the animation player to use this API to dosome more "complex" operations (like animation blending with IK, or IK solving on event emitted, etc..).

What you want to do is not so difficult right now but it's very uncomfortable because you have to take care of all things like event managing manually.

In the video that I've posted you can see that the IK solver is running only when the body is inside the button area, when it go outside the solver stop running. Here I could integrate it with an animation but as I said above I should take care of all event  managing manually.

mysticfall commented:
In Unity, there's `SetIKPositionWeight` and `SetIKRotationWeight` pair with which we can specify how much the IK would affect the animation.

If there was such a function in Godot, I thought it could be a lot easier to deal with such problems that I described above.

For instance, I might add an additional function track in the punching animation which can gradually change the IK weight/influence at the correct moment (i.e. just before the fist hits the target) to achieve the ideal result.

I was wondering if something like this could be implemented with the current binary on/off flag that is included in this PR. And if not, I think it'd be nice to have an API to change IK weight before we can implement more advanced features like the one you described.

Of course, I'm not saying that you should make the change right away, but I just wanted to know if I should open a feature request for it.

The video looks cool, by the way. Can't wait to try it myself :)

AndreaCatania commented:
The weight essentially is a parameter for blending animation. I'm imagine a scenario where you have an animation tree like this: 

![untitled](https://user-images.githubusercontent.com/8342599/39759699-a3077b7e-52d3-11e8-919c-57ee289e31c6.png)

For this reason I think that we don't have to manage it inside the IK sover, in this way we can take the two things separated and leave to the proper parts makes their job.

I think that manage IK inside the animation tree rather using the code is the best thing to do

mysticfall commented:
That makes sense. Then I'll wait until we overhaul `AnimationTreePlayer` before 3.1, hopefully with the IK integration feature you suggested. Thanks for the explanation!

AndreaCatania commented:
@akien-mga Sadly :|  `error: delegating constructors are permitted only in C++11`

Compilation error fixed
Code polished
Improved functions naming
Improved ik API workflow

(Updated "How it works" on first comment)

slapin commented:
About why I do need to keep some bones rigid in IK.

I have skeleton which I use for animation, it contains limbs of 4 bones, so 2 of bones act as one in case of IK (but that is not a case for animation - some poses simply require this to look ok, i.e. twisiting or muscule bulge simulation, small-scale bending). If IK will require all bones to be IKable, I will have to set up fake bones for IK and manually convert rotations to actual bones in frame loop which will eat-up CPU cycles. Swichig meshes/skeletons are also a problem (i.e. I want to use IK to fixup animation, not just posing, so I will need full skeleton for that).

AndreaCatania commented:
Talk about full skeleton IK is wrong. If you try this pr you will see that you have to create an IK task where you have to set root and tip bones. If the two bones to handle the muscle are outside of this chain its basis will not be touched.

When you create the task the chain is composed only of direct bones, all other appended bones doesn't partecipate to the solving and its basis is not touched 

 
![](https://i.imgur.com/61vMXRW.png)

If this is not your case please send me an image that represent your skeleton

slapin commented:
Well, the chain is constructed of 4 bones, [base]upperleg01.L,
upperleg02.L, [kneel is here] lowerleg01.L, lowerleg02.L[tip]
So in this case we need to implement IK over 2 bones - upperleg01 and
lowerleg01 but we neeed to consider whole chain length.


On Thu, May 10, 2018 at 5:45 PM, Andrea Catania <notifications@github.com>
wrote:

> Talk about full skeleton IK is wrong. If you try this pr you will see that
> you have to create an IK task where you have to set root and tip bones. If
> the two bones to handle the muscle are outside of this chain its basis will
> not be touched.
>
> When you create the task the chain is composed only of direct bones, all
> other appended bones doesn't partecipate to the solving and its basis is
> not touched
>
>
> <https://camo.githubusercontent.com/ae34e0b237bbe5a14a75894fc1e03644cd9d1452/68747470733a2f2f692e696d6775722e636f6d2f3631764d5852572e706e67>
>
> If this is not your case please send me an image that represent your
> skeleton
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-388075410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX05n-s3FbT-V-hg0MRz7vv1ZRGihjks5txFKYgaJpZM4T1-gF>
> .
>

LeonardMeagher2 commented:
@slapin are you saying the "kneel" is the parent to lowerleg01.L?
I'd imagine upperleg02.L should be a parent to both lowerleg01.L and "kneel"

slapin commented:
"Kneel" is not a bone, just the reference to understand the layout

Bones are: upperleg01.L upperleg02.L lowerleg01.L lowerleg02.L

upperleg01 is start of chain
lowerleg02 is end of chain

kneel is between upperleg02 and lowerleg01 which splits the limb into

upper part - upperleg01.L upperleg02.L
lower part - lowerleg01.L lowerleg02.L

each bening as single bone by IK and independently by animation.


On Thu, May 10, 2018 at 7:14 PM, Leonard Meagher <notifications@github.com>
wrote:

> @slapin <https://github.com/slapin> are you saying the "kneel" is the
> parent to lowerleg01.L?
> I'd imagine upperleg02.L should be a parent to both lowerleg01.L and
> "kneel"
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-388102931>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX0yVxbTxcVCPBXrVO7c73-DNRoE6Aks5txGdNgaJpZM4T1-gF>
> .
>

pwab commented:
As discussed with @TwistedTwigleg in godotengine/godot-demo-projects/issues/211 I also tried to impement FABRIK but for 2d IK chains. Is your implementation somehow transferable for 2d skeletons, @AndreaCatania?

AndreaCatania commented:
Yes it's possible, it's not too difficult take the code that I written and change Vector3 with Vector2 etc.. Feel free to do it if you want

Zireael07 commented:
Is this merge-ready or is it on hold like the soft body PR?

AndreaCatania commented:
It works, but reduz said that I need to add a function to blend with current animation pose. I think that I can do it today. I let you know

AndreaCatania commented:
I can't do it today I will finish it this weekend sorry

AndreaCatania commented:
I've implemented the blending functionality and this PR is ready to be merged.

Now is possible to use it with animation player so you can use IK easily with your animations.
Here a video: https://youtu.be/PXhF78c6wRs
and here the project to test it, and see how I've implemented it using animation player: 
[ik.zip](https://github.com/godotengine/godot/files/2086875/ik.zip)

AndreaCatania commented:
I've implemented the new node SkeletonIK as @reduz proposed.

Now is possible to set a **root** and **tip** bone and a target transform or a spatial node as target directly in the SkeletonIK node.
Then by calling the method **start** now the IK will be solved automatically and if you set a node as target the IK will follow it.

And also is possible to set the interpolation value, that can be changed at runtime, check this project to test it: [ik.zip](https://github.com/godotengine/godot/files/2108912/ik.zip)

Let me know what you think about!

AndreaCatania commented:
The branch is `AndreaCatania:ik`

> Also did not you consider trivial cases like 2 bones with target node and pole?

What do you mean?


> Also as was discussed before, was there any ideas regarding gaps in bone chain (for several bones acting as single bone)?

This is really particular case, and could be addressed later when this PR will be merged. A solution could be trivially use two separate SkeletonIK (chain) but should be tested in order to understand the use case and if the result is correct

slapin commented:
I mean 2 bone IK is special case and optimization. You can directly solve 2
bone IK and not run any algorithms like CCD/FABRIK, etc.
No iterations, just the result.
This allows to have complicated IK setups cheaply (and be simple for simple
set ups). I hope this is not too hard to understand.
The generic IK case it too slow for real practical (runtime) use,
especially for complex scenes. Most of practical set ups for aim (arms) and
uneven ground (feet) use 2 bone IK.

On Sun, Jun 17, 2018 at 6:40 PM, Andrea Catania <notifications@github.com>
wrote:

> I still don't understand correctly your concerns, anyway the IK chain can
> be maded by 1 to infinite bones, and both the root and tip of chain can be
> any bones
>
> So if you have for example a leg with a feet you can create the chain that
> start from the leg to the shin (2 bones). After solving the chain, the feet
> will be correctly placed according to the shin transform.
>
> The target that you set will affect the tip bone that in this case is the
> shin and also the feet because the basis where the feet is attached change.
>
> Sorry if I didn't understand what you said and my response is wrong but in
> this moment I can achieve the same behavior that I saw on the video without
> any problem and easily
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-397887144>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX090HF5rPCFAZRV-IiiWD9LeFsomSks5t9nh9gaJpZM4T1-gF>
> .
>

LeonardMeagher2 commented:
If you look up the law of cosines, it can solve 2 bone ik pretty easily

AndreaCatania commented:
Oh, you was talking about optimization... Now I see what you are talking about... Anyway I will see if it worth implement it because FABRIK is not so heavy algorithm expecially in case you have just two bones.

Do you have a prove of its bad performance insue that I can check?

reduz commented:
I'm on Andrea on this one, I don't think this is a case worth optimizing.

On Sun, Jun 17, 2018 at 4:58 PM Andrea Catania <notifications@github.com>
wrote:

> Oh, you was talking about optimization... Now I see what you are talking
> about... Anyway I will see if it worth implement it because FABRIK is not
> so heavy algorithm expecially in case you have just two bones.
>
> Do you have a prove of its bad performance insue that I can check?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-397902425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2yIhlKKSygTpvFTG2GKEm9xmqaysks5t9rTqgaJpZM4T1-gF>
> .
>

slapin commented:
Well, what is bad and good is basically a comparison. My target is 200
characters per scene.
I did not try your implementation, but with some other implementations I
find that I can have
5 characters per scene with CCD and 20 with FABRIK (minus heavy animation
stuff which sucks).
With 2-bone IK I can have 200 characters on the same hardware (running
animation).
Of cource LODs and animation lods and manual LODs are utilized at the same
extent on all the scenes.


On Sun, Jun 17, 2018 at 10:58 PM, Andrea Catania <notifications@github.com>
wrote:

> Oh, you was talking about optimization... Now I see what you are talking
> about... Anyway I will see if it worth implement it because FABRIK is not
> so heavy algorithm expecially in case you have just two bones.
>
> Do you have a prove of its bad performance insue that I can check?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-397902425>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX05NqyoSs3x6fTdAUmfrEAeXVNXV7ks5t9rT4gaJpZM4T1-gF>
> .
>

slapin commented:
with FABRIK you will need a few iterations in best case, while with 2-bone
IK you will
give direct single answer without iterations (and that is 5 lines of code).

On Mon, Jun 18, 2018 at 2:36 AM, Sergey Lapin <slapinid@gmail.com> wrote:

> Well, what is bad and good is basically a comparison. My target is 200
> characters per scene.
> I did not try your implementation, but with some other implementations I
> find that I can have
> 5 characters per scene with CCD and 20 with FABRIK (minus heavy animation
> stuff which sucks).
> With 2-bone IK I can have 200 characters on the same hardware (running
> animation).
> Of cource LODs and animation lods and manual LODs are utilized at the same
> extent on all the scenes.
>
>
> On Sun, Jun 17, 2018 at 10:58 PM, Andrea Catania <notifications@github.com
> > wrote:
>
>> Oh, you was talking about optimization... Now I see what you are talking
>> about... Anyway I will see if it worth implement it because FABRIK is not
>> so heavy algorithm expecially in case you have just two bones.
>>
>> Do you have a prove of its bad performance insue that I can check?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/18699#issuecomment-397902425>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAAX05NqyoSs3x6fTdAUmfrEAeXVNXV7ks5t9rT4gaJpZM4T1-gF>
>> .
>>
>
>

AndreaCatania commented:
In case of 200 characters I think that for 90% of these you don't need IK at all, anyway I will add some functions in order to expose max iterations and bias so you can control how precision should be.

Also is interesting for me test this scene with 200 character to see what we are talking about

slapin commented:
For me the most common scenario where I need IK + 200 people is
like
https://www.pixoto.com/images-photography/people/street-and-candids/crowded-spanish-steps-in-rome-5019323078279168
where errors
in feet motion are visible and at the same time you don't want to make it
fixed - that will be easily noticed as repeatitive, so procedural
animation is required.
As Ubisoft can do this I think Godot should too...
https://www.youtube.com/watch?v=kOv5ixy7Y8w

Anyway ease of LOD usage with IK and animations should be required to do
true massive crowds like in Ubisoft games.
(they also replace distant characters with much simpler meshes with much
simpler skeletons, but that is another story).


On Mon, Jun 18, 2018 at 9:01 AM, Andrea Catania <notifications@github.com>
wrote:

> In case of 200 characters I think that for 90% of these you don't need IK
> at all, anyway I will add some functions in order to expose max iterations
> and bias so you can control how precision should be.
>
> Also is interesting for me test this scene with 200 character to see what
> we are talking about
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-397949974>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAX03ozYvBH-62ZiP0n_H8xQfpCbb4wks5t90IlgaJpZM4T1-gF>
> .
>

AndreaCatania commented:
In this picture the error on the feet are visible only on the first line of people sitdown on the wall, and for all others you don't need the IK to fix it.

The same rule goes with Ubisoft video, do you think that all people that are so faraway that seems just a texture have the IK active? I don't think so... you need a perfect IK only for nearest person and should be activated only when required so when a person is out of focus but still near and visible you need to solve IK with just a good precision, but for all other you don't need to solve..

Also when you resolve the IK one time you don't need to re solve it again if you don't move the bones.

Anyway I think that for talk about what we have to change to optimize we first need a scene that runs at low FPS to check it... because in this moment we are talk about a real picture and what you could need to make it in godot

AndreaCatania commented:
The IK had the problem that when you wanted to make it works with the animation player, the animation player override the bone poistion if executed after IK solver.

In order to make it works correctly I've moved the execution code from process_internal to process, so I'm always sure that the IK solver is processed always after all animation players.

This doesn't make any differences on Editor side and is always possible to override the _process function by a script.

hpvb commented:
@andreacatania are you happy with the current state of this PR? I think it should be merged if you are happy.

AndreaCatania commented:
No is required to make a little change that can't be done without merging
this pr : #19888

On Sat, Jul 14, 2018, 13:29 Hein-Pieter van Braam <notifications@github.com>
wrote:

> @AndreaCatania <https://github.com/AndreaCatania> are you happy with the
> current state of this PR? I think it should be merged if you are happy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/18699#issuecomment-405017097>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH9MR6w_onNbHZvPX5rURJFB281J0Ze5ks5uGdYDgaJpZM4T1-gF>
> .
>

karroffel commented:
I mentioned this before in a private discussion, but I'll leave it here so others can see it too.

I tested this PR, and while it works really well, the interface and way to use it is quite different than in pretty much all other software that let's the user use IK.

With this PR, a target transform is applied to the bone on the tip of the chain, then the rest of the chain "follows". That means that the *rotation* of the target has to match the one of the desired pose.

This is a lot harder to get right.

The usual approach is to have a target (that only contributes position by default, rotation opt-in) and a *pole* that dictates how the chain will bend.

From what I have seen, there is an official IK demo that adds a "middle bone helper" (so a pole basically) on top of the FABRIK algorithm.

https://github.com/godotengine/godot-demo-projects/blob/c6b481aaf0cfb677b182b0e3711702c15b0cb413/3d/ik/addons/sade/ik_fabrik.gd

Maybe this PR can be modified so it supports the more "standard" target+pole workflow.

reduz commented on scene/animation/skeleton_ik.h:
```diff
@@ -0,0 +1,202 @@
+class SkeletonIK : public Node {
+	GDCLASS(SkeletonIK, Node);
+
+	String root_bone;
```
by the way, I think this should be StringName, not String

AndreaCatania replied:
I changed it, but seems to me that it is not better, because the skeleton internally store bone names as String

AndreaCatania replied:
I've tried to set StringName anyway, but I got compilation error..

reduz commented on scene/animation/skeleton_ik.h:
```diff
@@ -0,0 +1,202 @@
+	void set_target(const Transform &p_target);
+	const Transform &get_target() const;
+
+	void set_target_node_override(const NodePath &p_node);
```
I dont think this should be called override, if set it should be obvious (And docummented) that takes priority over a transform

AndreaCatania replied:
Done

reduz commented on scene/animation/skeleton_ik.h:
```diff
@@ -0,0 +1,202 @@
+	void set_interpolation(real_t p_interpolation);
+	real_t get_interpolation() const;
+
+	void set_target(const Transform &p_target);
```
I would call this set_target_transform

AndreaCatania replied:
done

reduz commented on scene/animation/skeleton_ik.h:
```diff
@@ -0,0 +1,202 @@
+	NodePath get_target_node_override();
+
+	void set_min_distance(real_t p_min_distance);
+	real_t get_min_distance() const { return min_distance; }
```
I dont quite understand what this is for

AndreaCatania replied:
This is used to control the precision of IK, it tells the minimum distance between target and tip before consider the process done.

AndreaCatania commented:
@reduz  I've a problem with process_internal priority.

The execution order seems random, (or better seems related to the area of last mouse click), when I start the animation and IK in the editor the IK sometimes is processed before the animation player and sometimes later...
Even if the priority of AnimatorPlayer is 0 and should be executed before SkeletonIK that has the priority 1.

check video: https://www.youtube.com/watch?v=yBliluCY6M0 

Ther's something other that I need to address to properly implement the priority?

AndreaCatania committed:
Fixed process priority

akien-mga commented:
Needs a rebase to fix conflicts in skeleton.cpp.

AndreaCatania commented:
Rebased

reduz commented on scene/main/node.h:
```diff
@@ -72,7 +72,7 @@ class Node : public Object {
 
 	struct ComparatorWithPriority {
 
-		bool operator()(const Node *p_a, const Node *p_b) const { return p_b->has_priority_higher_than(p_a) || p_b->is_greater_than(p_a); }
```
Just request the priority (integer) and compare integers here

reduz replied:
no need to add a function that no one else will use
